spock-specs/src/test/groovy/org/spockframework/smoke/extension/TempDirExtensionSpec.groovy

```diff
@@ -0,0 +1,113 @@
+package org.spockframework.smoke.extension
+
+
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Stepwise
+import spock.lang.TempDir
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+
+/**
+ * @author dqyuan
+ */
+@Stepwise
+class TempDirExtensionSpec extends Specification {
+
+  @Shared
+  @TempDir
+  Path sharedDir
+  @Shared
+  Path previousShared
+
+  @TempDir
+  File iterationDir
+  @Shared
+  File previousIteration
+
+  @TempDir(baseDir = "build")
+  Path customerTempDir
+
+  @TempDir(baseDir = "build", keep = { true })
+  Path reserveDir
+
+  @Shared
+  Path reserveFile
+
+  @TempDir
+  def untypedPath
+
+  def "temp dir exist"() {
+    previousShared = sharedDir
+    previousIteration = iterationDir
+    reserveFile = reserveDir.resolve("test.txt")
+    Files.write(reserveFile, "aaa".getBytes())
+
+    expect:
+    Files.exists(sharedDir)
+    iterationDir.exists()
+    iterationDir.directory
+    Files.exists(customerTempDir)
+    Files.exists(reserveDir)
+    untypedPath instanceof Path
+    customerTempDir.parent.fileName.toString() == 'build'
+    reserveDir.parent.fileName.toString() == 'build'
+  }
+
+  def "reserve temp directory when reserveAfterTest is true"() {
+    expect:
+    Files.exists(reserveFile)
+  }
+
+  def "@TempDir creates only one dir for one spec, if the annotated field is 'shared'"() {
+    expect:
+    sharedDir == previousShared
+  }
+
+  def "get a different path for non-shared field, and previous was deleted"() {
+    expect:
+    iterationDir != previousIteration
+    !previousIteration.exists()
+  }
+
+  def "@TempDir creates one temp directory per iteration for normal field"(String foo, int bar) {
+    expect:
+    iterationDir != previousIteration
+    foo == "fo"
+    bar == 1
```

Comment:
do the foo and bar checks actually have any value for the test?
I asked this before and also suggested to replace the data table by a simple pipe with int-range.
I really wonder what the value for the test is of these. :-)

Hm, it seems actually that GitHub somehow lost the comment about the datapipe. :-/
What I suggested was this (you do not need to declare the data variables as parameters, even when you use them in the body):

```groovy
  def "@TempDir creates one temp directory per iteration for normal field"() {
    expect:
    iterationDir != previousIteration

    cleanup:
    previousIteration = iterationDir

    where:
    i << (1..3)
  }
```